README.md
# loopback-connector-couchdb2

The `loopback-connector-couchdb2` module is the CouchDB 2.0 connector for the
LoopBack framework that supports the advanced functionality originally found
only in Cloudant but that is now available in CouchDB.

<!-- TOC -->

- [Getting Started](#getting-started)
  - [Model](#model)
    - [Map Between Model And Document](#map-between-model-and-document)
    - [Model-specific Configuration](#model-specific-configuration)
    - [_rev Property](#_rev-property)
      - [Example CRUD operations with `_rev`](#example-crud-operations-with-_rev)
- [Setup Couchdb Instance](#setup-couchdb-instance)
- [Installation](#installation)
- [Configuration](#configuration)
  - [Generate Datasource](#generate-datasource)
  - [Datasource Config](#datasource-config)
  - [Example Usage](#example-usage)
- [CRUD](#crud)
- [Migration](#migration)
  - [autoupdate vs automigrate](#autoupdate-vs-automigrate)
  - [isActual](#isactual)
  - [property index](#property-index)
  - [Example Code](#example-code)
- [Discovery](#discovery)
- [Query](#query)
- [View](#view)
- [Bulk replace](#bulk-replace)
- [Testing](#testing)
  - [Docker](#docker)
- [More into](#more-info)

<!-- /TOC -->

# Getting Started

## Model 

### Map Between Model And Document

Similar to Cloudant, Couchdb doesn't have a concept as 'table' or 'collection', and to support ad-hoc query which is an important loopback feature, by default the connector uses all_fields index for query, and doesn't create design document for a loopback model. 

In a real application, all_fields index doesn't do any optimization and slow down the performance a lot, for details about how to create index for properties, please refer to [property index](https://github.com/strongloop/loopback-connector-couchdb2#property-index)

A loopback model instance is stored as a document in Couchdb. It has a model index property to specify the model name, the connector also adds it to Couchdb query's selector when doing model level queries. For example, a User model instance is stored as

```
"loopback__model__name": "User",
"username": "Foo",
"password": "bar"
```

To create a model instance, the connector creates a document with value of property 'loopback__model__name' equals to `User`, and adds `loopback__model__name: 'User'` to query when fetches `User` instances.

By default, `modelIndex` is 'loopback__model__name', and `modelSelector` is {[modelIndex]: modelName}. User can customize `modelSelector` or `modelIndex` in model's json file. For details please check [model-specific configuration](https://github.com/strongloop/loopback-connector-couchdb2#model-specific-configuration)

### Model-specific Configuration

You can specify configurations per model for database selection and to
map a model to a different document:

*common/models/_model-name_.json*

```
{
  "name": "User",
  "base": "PersistedModel",
  "idInjection": true,
  ...
  "couchdb": {
    "modelIndex": "custom_model_index_name",
    "modelSelector": { "custom_selector": "user" },
    "database": "test2"
  },
  ...
```

Model-specific configuration settings:

Property&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  | Type | Description
----------| -----| --------
database  | String | Database name
modelIndex | String | Specify the model name to document mapping, defaults to `loopback__model__name`.
modelSelector | JSON | Use the Couchdb Query selector syntax to associate models to existing data. NOTE: modelSelector and modelIndex are mutually exclusive. modelSelector will override modelIndex when building query.

### _rev Property

In a document, property `_rev` is the latest doc revision and must be provided when modifying the doc.

Our connector allows the user to retrieve back the `_rev` property upon all CRUD operations, however does not add it to the model definition.

If you would like to have a `_rev` property on your model, as an end user, the onus is on you to add the property in the model definition.

**Note:** All CRUD operations require `_rev` (except create) and __it is up to the user to specify them__. The connector does not handle such cases due to possibilities of race condition when two users try to update the same document.

#### Example CRUD operations with `_rev`

`model.json`
``` json
{
  ... 
  "properties": {
    "_rev": {
      "type": "string"
    },
    "name": {
      "type": "string"
    }
  },
  ...
}
```

- Create

```javascript
  Model.create([{
    name: 'Foo',
  }, {
    name: 'Bar',
  }], function(err, result) {
    if (err) throw err;
    console.log('Created instance: ' + JSON.stringify(result));
  });
```

**Note:** Couchdb does not allow customized `_rev` value, hence creating an instance with a `_rev` value will not give the expected result (i.e Couchdb's CREATE operation ignores the `_rev` value when provided and generates a random unique one). The onus is on the user if they fail to comply to this rule.

Let's say we have an instance in the database:
```json
{
   "id":"2",
   "_rev":"2-abcedf",
   "name":"Bar"
}
```

- Find

  - find

  ```javascript
    Model.find(function(err, result) {
      if (err) throw err;
      console.log('Found all instances: ' + JSON.stringify(result));
    });
  ```

  - findById

  ```javascript
    Model.findById('2', function(err, result) {
      if (err) throw err;
      console.log('Found instance with id: ' + JSON.stringify(result));
    });
  ```

- Replace

  - replaceOrCreate

  ```javascript
    Model.replaceOrCreate({
      id:'2',
      _rev:'2-abcedf',
      name:'Bar2'
    }, function(err, result) {
      if (err) throw err;
      console.log('Replace an existing instance: ' + JSON.stringify(result));
    });
  ```

  - replaceById

  ```javascript
    Model.replaceById('2', {
      _rev:'2-abcedf',
      name:'Bar3'
    }, function(err, result) {
      if (err) throw err;
      console.log('Replace an existing instance with id: ' + JSON.stringify(result));
    });
  ```

- Update

  - updateOrCreate

  ```javascript
    Model.updateOrCreate({
      id:'2',
      _rev:'2-abcedf',
      name:'Bar4'
    }, function(err, result) {
      if (err) throw err;
      console.log('Update an existing instance: ' + JSON.stringify(result));
    });
  ```

  - update/updateAll

    - with `_rev` property
      ```javascript
        Model.updateAll({
          _rev:'2-abcedf',
          name:'Bar4'
        }, {name: 'Bar4-updated', _rev: '2-abcedf'}, function(err, result) {
          if (err) throw err;
          console.log('Update an existing instance: ' + JSON.stringify(result));
        });
      ```

    - without `_rev` property
      ```javascript
        Model.updateAll({
          name:'Bar4'
        }, {name: 'Bar4-updated'}, function(err, result) {
          if (err) throw err;
          console.log('Update an existing instance: ' + JSON.stringify(result));
        });
      ```

# Setup Couchdb Instance

For users that don't have a Couchdb server to develop or test, here are some suggestions can help you quickly set one up.

For development use, a docker container is easy to setup. Users can also download the on-prem Couchdb2.0 from http://couchdb.apache.org/

# Installation

Enter the following in the top-level directory of your LoopBack application:

```
$ npm install loopback-connector-couchdb2 --save
```

The `--save` option adds the dependency to the application’s `package.json` file.

# Configuration

## Generate Datasource

Use the [Data source generator](http://loopback.io/doc/en/lb3/Data-source-generator.html) to add the Couchdb data source to your application. The entry in the applications `/server/datasources.json` will
look something like this:

```
"mydb": {
  "name": "mydb",
  "connector": "couchdb",
  "url": "https://<username>:<password>@<host>"
  "database": "test"
}
```

## Datasource Config
The connector passes all configurations to nano driver, please check couchdb-nano's document for details:
https://github.com/apache/couchdb-nano#configuration

## Example Usage

*/server/script.js*
```javascript
var util = require('util');

// Here we create datasource dynamically.
var DataSource = require ('loopback-datasource-juggler').DataSource,
    Couchdb = require ('loopback-connector-couchdb2');

var config = {
    url: 'your_couchdb_url'
    database: 'your_couchdb_database'
};

var db = new DataSource (Couchdb, config);

Test = db.define ('Test', {
  name: { type: String },
});

Test.create({
  name: "Tony",
}).then(function(test) {
  console.log('create instance ' + util.inspect(test, 4));
  return Test.find({ where: { name: "Tony" }});
}).then(function(test) {
  console.log('find instance: ' + util.inspect(test, 4));
  return Test.destroyAll();
}).then(function(test) {
  console.log('destroy instance!');
}).catch(err);

});
```

- Use different DB instances per model definition. Refer to https://github.com/strongloop/loopback-connector-couchdb2/blob/master/doc/multiple-db-instances.md

# CRUD

User can find most loopback CRUD operation apis documented in https://loopback.io/doc/en/lb3/Built-in-models-REST-API.html

Due to the `_rev` property, Couchdb connector handles CRUD functions a little differently, for details and examples please refer to [_rev-property](https://github.com/strongloop/loopback-connector-couchdb2/blob/master/doc/_rev-property.md)

# Migration

For a model connected to Couchdb database, migration means create/update a design document with proper indexes provided by the model. There is a section called [property index](https://github.com/strongloop/loopback-connector-couchdb2#property-index) that talks about how to define indexes. 

After attaching a model to a Couchdb datasource, either statically with `model.json` file or dynamically in boot script code, user need to run `automigrate` or `autoupdate` to migrate models to database. Couchdb connector does NOT automatically migrate them.

The following migration functions take either an array of multiple model's name, or a string of a single model's name. The example code will show how to do it.

## autoupdate vs automigrate

`autoupdate` does not destroy existing model instances if model already defined in database. It only creates design document for new models. 
Under the hood Couchdb allows creating same design doc multiple times, it doesn't return error, but returns `existed` as result to tell is it a new design doc or existing one.

`automigrate` destroys existing model instances if model already defined in database. Please make sure you do want to clean up data before running `automigrate`. Then it does same thing as `autoupdate`

## isActual

User can call this function to check if model exists in database.
We need to discuss do we still want to create a design doc for a model if no index provided:
- if yes: keep this function
- if no: isActual doesn't make any sense then

## property index

TBD. Briefly:
- By default we use all_fields index with no optimization for performance
- If user define indexable properties or composite index, we create them in one design document
- It's upon user's choice to specify the index they want to use in a query.

## Example Code

Should be adjusted according to the decision we made for isActual

*/server/script.js*
```javascript
module.export = function migrateData(app) {
  // Suppose you already define a datasource called `cloudantDS` 
  // in server/datasources.json
  var ds = app.datasources.cloudantDS;
  
  // static model created with model.json file
  var StaticModel = app.models.StaticModel;
  // dynamic model created in boot script
  var DynamicModel = ds.define('DynamicModel', {
      name: {type: String},
      description: {type: String},
  });

  // Write the three examples in parallel just to avoid dup code,
  // please try ONLY ONE of them at one time.
  ds.once('connected', function() {
    // try autoupdate example - multiple models
    ds.autoupdate(['StaticModel', 'DynamicModel'], function(err) {});
    // OR
    // try automigrate example - single model
    ds.automigrate('StaticModel', function(err) {});
    // OR
    // try isActual example - if any model exist, run autoupdate, otherwise automigrate
    ds.isActual(['StaticModel', 'DynamicModel'], function(err, exist) {
      if (exist) {
        ds.autoupdate(['StaticModel', 'DynamicModel'], function(err){})
      } else {
        ds.automigate(['StaticModel', 'DynamicModel'], function(err){});
      }
    });
  });
}
```

# Discovery

Not implemented yet in this connector.

# Query

-  Couchdb doesn't support sorting with a property that's not indexable.
-  [LoopBack query](http://loopback.io/doc/en/lb3/Querying-data.html) support for: fields, limit, order, skip and where filters.
- Please check [Advanced Queries](https://github.com/strongloop/loopback-connector-couchdb/blob/master/doc/advanced-queries.md) for details about regex filter, nested filter and order.

# View

Given a design doc name and the view name in it, user can use a connector level function `viewDocs` to query the view.

Since `viewDocs` is a specific api for Couchdb/Cloudant connector only, it is not attached to the dataSource Object defined in loopback-datasource-juggler, which means the correct way to call it is `ds.connector.viewDocs`:

*/server/script.js*
```javascript
module.exports = function(server) {
  // Get Couchdb dataSource as `ds`
  // 'couchdbDS' is the name of Couchdb datasource created in 
  // 'server/datasources.json' file
  var ds = server.datasources.couchdbDS;

  // 1. Please note `ds.connector.viewDocs()` is the correct way to call it,
  // NOT `ds.viewDocs()`
  // 2. This api matches the Couchdb endpoint:
  // GET /db/_design/<design-doc>/_view/<view-name>
  ds.connector.viewDocs('design_doc', 'view_name', function(err, results) {
    // `results` would be the data returned by quering that view
  });

  // Alternatively user can also specify the filter for view query
  ds.connector.viewDocs('design_doc', 'view_name', {key: 'filter'}, 
    function(err, results) {});
};
```

# Bulk replace

Given an array of data to be updated, Couchdb supports the idea of performing bulk replace on a model instance. Please note, unlike other CRUD operations, bulk replace does not invoke any operation hooks.

**Note:** To perform bulk replace, each data in the array data set needs to have the `id` and `_rev` property corresponding to the documents `id` and `_rev` property in the database.

Example:

```server/boot/script.js```

```javascript
var dataToCreate = [
  {id: 1, name: 'Foo', age: 1},
  {id: 2, name: 'Bar', age: 1},
  {id: 3, name: 'Baz', age: 2},
  {id: 4, name: 'A', age: 4},
  {id: 5, name: 'B', age: 5},
  {id: 6, name: 'C', age: 6},
  {id: 7, name: 'D', age: 7},
  {id: 8, name: 'E', age: 8},
 ];
var dataToUpdate = [
 {id: 1, name: 'Foo-change', age: 11},
 {id: 5, name: 'B-change', age: 51},
 {id: 8, name: 'E-change', age: 91}
];

module.exports = function(app) {
  var db = app.dataSources.couchdbDS;
  var Employee = app.models.Employee;

  db.automigrate(function(err) {
    if (err) throw err;

    Employee.create(dataToCreate, function(err, result) {
    if (err) throw err;
    console.log('\nCreated instance: ' + JSON.stringify(result));

    dataToUpdate[0].id = result[0].id;
    dataToUpdate[0]._rev = result[0]._rev;
    dataToUpdate[1].id = result[4].id;
    dataToUpdate[1]._rev = result[4]._rev;
    dataToUpdate[2].id = result[7].id;
    dataToUpdate[2]._rev = result[7]._rev;

    // note: it is called `db.connector.bulkReplace`
    // rather than `Employee.bulkReplace`
    db.connector.bulkReplace('Employee', dataToUpdate, function(err, result) {
        if (err) throw err;

        console.log('\nBulk replace performed: ' + JSON.stringify(result));

        Employee.find(function(err, result) {
        if (err) throw err;

        console.log('\nFound all instances: ' + JSON.stringify(result));
        });
    });
  });
});
};
```

# Testing

## Docker
- Assuming you have [Docker](https://docs.docker.com/engine/installation/) installed, run the following script which would spawn a Couch instance on your local:
```bash
source setup.sh <HOST> <USER> <PASSWORD> <PORT> <DATABASE>
```
 where `<HOST>`, `<PORT>`, `<USER>`, `<PASSWORD>` and `<DATABASE>` are optional parameters. The default values are `localhost`, `5984`, `admin`, `pass` and `testdb` respectively.
- Run the test:
```bash
npm run mocha
```

# More Info
For more detailed information regarding connector-specific functions and behaviour,
see the [docs section](https://github.com/strongloop/loopback-connector-couchdb2/tree/master/doc).


CHANGES.md
2019-11-26, Version 1.5.3
=========================

 * feat: add partition field to model (#72) (Janny)

 * chore: improve issue and PR templates (Nora)


2019-11-19, Version 1.5.2
=========================

 * refactor: add options for search (#71) (Janny)


2019-11-14, Version 1.5.1
=========================

 * refactor: migrate functions (#69) (Janny)


2019-09-19, Version 1.5.0
=========================

 * drop support for node 6 (Nora)


2019-05-07, Version 1.4.0
=========================

 * chore: update strong-globalize version (Diana Lau)

 * chore: update copyrights years (Diana Lau)

 * fix: update lodash (#58) (Janny)

 * fix: honor fields in the filter (#59) (Janny)

 * fix: eslint (#56) (Janny)


2018-07-09, Version 1.3.0
=========================

 * fix: lint (#55) (Janny)

 * chore: fix linting and add .npmrc (virkt25)

 * PCRE Regular Expressions (#52) (Dan Jarvis)

 * [WebFM] cs/pl/ru translation (#50) (tangyinb)

 * remove node 4 (jannyHou)

 * correct the connector name (jannyHou)


2018-04-27, Version 1.2.1
=========================

 * fix: add model view in index (#47) (Janny)


2018-01-19, Version 1.2.0
=========================

 * Add global limit (#42) (Janny)

 * Fix lint (#38) (Janny)


2017-10-18, Version 1.1.0
=========================

 * update dependencies (Diana Lau)


2017-09-26, Version 1.0.3
=========================

 * Update nano version (#36) (Loay)

 * remove event listener (#33) (Janny)

 * Add stalebot configuration (Kevin Delisle)


2017-08-22, Version 1.0.2
=========================

 * Recover skipped test (ssh24)


2017-08-21, Version 1.0.1
=========================

 * getMoSettings (#30) (Janny)

 * Make getDriver extendable (#29) (Janny)

 * Fix readme (ssh24)


2017-08-17, Version 1.0.0
=========================

 * Remove beta warning (Kevin Delisle)

 * Using strong-mocha-interfaces instead of lb-bdd (#25) (Taranveer Virk)

 * Implement indexes feature (ssh24)

 * Create Issue and PR Templates (#24) (Sakib Hasan)

 * Fix npm mocha command (ssh24)

 * Use Mocha Interface to skip tests (Taranveer Virk)

 * Update eslint (ssh24)

 * Add test folder (#17) (Janny)

 * Return data without model name prop (ssh24)

 * Add CODEOWNER file (Diana Lau)


2017-08-07, Version 0.9.0
=========================

 * First release!


test.js
// Copyright IBM Corp. 2017,2019. All Rights Reserved.
// Node module: loopback-connector-couchdb2
// This file is licensed under the Apache License 2.0.
// License text available at https://opensource.org/licenses/Apache-2.0

'use strict';

var _ = require('lodash');
var async = require('async');
var spawn = require('child_process').spawn;
var docker = new require('dockerode')();
var fmt = require('util').format;
var http = require('http');
var ms = require('ms');

// we don't pass any node flags, so we can call _mocha instead the wrapper
var mochaBin = require.resolve('mocha/bin/_mocha');

process.env.COUCHDB_DATABASE = 'test-db';
process.env.COUCHDB_PASSWORD = 'pass';
process.env.COUCHDB_USERNAME = 'admin';
// process.env.COUCHDB_DATABASE_CREATE = true;

// these are placeholders. They get set dynamically based on what IP and port
// get assigned by docker.
process.env.COUCHDB_PORT = 'TBD';
process.env.COUCHDB_HOST = 'TBD';
process.env.COUCHDB_URL = 'TBD';

var CONNECT_RETRIES = 30;
var CONNECT_DELAY = ms('5s');

var containerToDelete = null;

var testOpts = [mochaBin, '--timeout', '40000', '--require', 'strong-mocha-interfaces', '--require', 'test/init.js', '--ui', 'strong-bdd'];
if (process.env.MOCHA_OPT_GREP) {
  testOpts.push('--grep');
  testOpts.push(process.env.MOCHA_OPT_GREP);
}

async.waterfall([
  dockerStart('klaemo/couchdb:2.0.0'),
  sleep(ms('2s')),
  setCouchDBEnv,
  waitFor('/_all_dbs'),
  createDB('test-db'),
  run(testOpts),
], function(testErr) {
  dockerCleanup(function(cleanupErr) {
    if (cleanupErr) {
      console.error('error cleaning up:', cleanupErr);
    }
    if (testErr) {
      console.error('error running tests:', testErr);
      process.exit(1);
    }
  });
});

function sleep(n) {
  return function delayedPassThrough() {
    var args = [].slice.call(arguments);
    // last argument is the callback
    var next = args.pop();
    // prepend `null` to indicate no error
    args.unshift(null);
    setTimeout(function() {
      next.apply(null, args);
    }, n);
  };
}

function dockerStart(imgName) {
  return function pullAndStart(next) {
    console.log('pulling image: %s', imgName);
    docker.pull(imgName, function(err, stream) {
      docker.modem.followProgress(stream, function(err, output) {
        if (err) {
          return next(err);
        }
        console.log('starting container from image: %s', imgName);
        docker.createContainer({
          Image: imgName,
          HostConfig: {
            PublishAllPorts: true,
          },
          Env: [
            'COUCHDB_USER=' + process.env.COUCHDB_USERNAME,
            'COUCHDB_PASSWORD=' + process.env.COUCHDB_PASSWORD,
          ],
        }, function(err, container) {
          console.log('recording container for later cleanup: ', container.id);
          containerToDelete = container;
          if (err) {
            return next(err);
          }
          container.start(function(err, data) {
            next(err, container);
          });
        });
      });
    });
  };
}

function setCouchDBEnv(container, next) {
  container.inspect(function(err, c) {
    // if swarm, Node.Ip will be set to actual node's IP
    // if not swarm, but remote docker, use docker host's IP
    // if local docker, use localhost
    var host = _.get(c, 'Node.IP', _.get(docker, 'modem.host', '127.0.0.1'));
    // container's port 80 is dynamically mapped to an external port
    var port = _.get(c,
      ['NetworkSettings', 'Ports', '5984/tcp', '0', 'HostPort']);

    process.env.COUCHDB_PORT = port;
    process.env.COUCHDB_HOST = host;
    var usr = process.env.COUCHDB_USERNAME;
    var pass = process.env.COUCHDB_PASSWORD;
    process.env.COUCHDB_URL = 'http://' + usr + ':' + pass + '@' +
      host + ':' + port;
    console.log('env:', _.pick(process.env, [
      'COUCHDB_URL',
      'COUCHDB_HOST',
      'COUCHDB_PORT',
      'COUCHDB_USERNAME',
      'COUCHDB_PASSWORD',
      'COUCHDB_DATABASE',
      'COUCHDB_DATABASE_CREATE',
    ]));
    next(null, container);
  });
}

function waitFor(path) {
  return function waitForPath(container, next) {
    var opts = {
      host: process.env.COUCHDB_HOST,
      port: process.env.COUCHDB_PORT,
      auth: process.env.COUCHDB_USERNAME + ':' + process.env.COUCHDB_PASSWORD,
      path: path,
    };

    console.log('waiting for instance to respond');
    return ping(null, CONNECT_RETRIES);

    function ping(err, tries) {
      console.log('ping (%d/%d)', CONNECT_RETRIES - tries, CONNECT_RETRIES);
      if (tries < 1) {
        next(err || new Error('failed to contact CouchDB'));
      }
      http.get(opts, function(res) {
        res.pipe(devNull());
        res.on('error', tryAgain);
        res.on('end', function() {
          if (res.statusCode === 200) {
            setImmediate(next, null, container);
          } else {
            tryAgain();
          }
        });
      }).on('error', tryAgain);
      function tryAgain(err) {
        setTimeout(ping, CONNECT_DELAY, err, tries - 1);
      }
    }
  };
}

function createDB(db) {
  return function create(container, next) {
    var opts = {
      method: 'PUT',
      path: '/' + db,
      host: process.env.COUCHDB_HOST,
      port: process.env.COUCHDB_PORT,
      auth: process.env.COUCHDB_USERNAME + ':' + process.env.COUCHDB_PASSWORD,
    };
    console.log('creating db: %j', db);
    http.request(opts, function(res) {
      res.pipe(devNull());
      res.on('error', next);
      res.on('end', function() {
        setImmediate(next, null, container);
      });
    })
      .on('error', next)
      .end();
  };
}

function run(cmd) {
  return function spawnNode(container, next) {
    spawn(process.execPath, cmd, {stdio: 'inherit'})
      .on('error', next)
      .on('exit', onExit);

    function onExit(code, sig) {
      if (code) {
        next(new Error(fmt('mocha exited with code: %j, sig: %j', code, sig)));
      } else {
        next();
      }
    }
  };
}

// clean up any previous containers
function dockerCleanup(next) {
  if (containerToDelete) {
    console.log('cleaning up container: %s', containerToDelete.id);
    containerToDelete.remove({force: true}, function(err) {
      next(err);
    });
  } else {
    setImmediate(next);
  }
}

// A Writable Stream that just consumes a stream. Useful for draining readable
// streams so that they 'end' properly, like sometimes-empty http responses.
function devNull() {
  return new require('stream').Writable({
    write: function(_chunk, _encoding, cb) {
      return cb(null);
    },
  });
}


test/init.js
// Copyright IBM Corp. 2017,2019. All Rights Reserved.
// Node module: loopback-connector-couchdb2
// This file is licensed under the Apache License 2.0.
// License text available at https://opensource.org/licenses/Apache-2.0

'use strict';

module.exports = require('should');

var DataSource = require('loopback-datasource-juggler').DataSource;
var _ = require('lodash');

var config = {
  url: process.env.COUCHDB_URL,
  username: process.env.COUCHDB_USERNAME,
  password: process.env.COUCHDB_PASSWORD,
  database: process.env.COUCHDB_DATABASE,
  port: process.env.COUCHDB_PORT,
  plugin: 'retry',
  retryAttempts: 10,
  retryTimeout: 50,
  databaseCreate: process.env.COUCHDB_DATABASE_CREATE,
};

console.log('env config ', config);

global.config = config;
global.IMPORTED_TEST = false;

var skips = [
  'find all limt ten',
  'find all skip ten limit ten',
  'find all skip two hundred',
  'isActual',
];

if (process.env.LOOPBACK_MOCHA_SKIPS) {
  process.env.LOOPBACK_MOCHA_SKIPS =
    JSON.stringify(JSON.parse(process.env.LOOPBACK_MOCHA_SKIPS).concat(skips));
} else {
  process.env.LOOPBACK_MOCHA_SKIPS = JSON.stringify(skips);
}

global.getDataSource = global.getSchema = function(customConfig) {
  var db = new DataSource(require('../'), customConfig || config);
  db.log = function(a) {
    console.log(a);
  };

  var originalConnector = _.clone(db.connector);
  var overrideConnector = {};

  overrideConnector.save = function(model, data, options, cb) {
    if (!global.IMPORTED_TEST) {
      return originalConnector.save(model, data, options, cb);
    } else {
      var self = this;
      var idName = self.idName(model);
      var id = data[idName];
      var mo = self.selectModel(model);
      data[idName] = id.toString();

      mo.db.get(id, function(err, doc) {
        if (err) return cb(err);
        data._rev = doc._rev;
        var saveHandler = function(err, id) {
          if (err) return cb(err);
          mo.db.get(id, function(err, doc) {
            if (err) return cb(err);
            cb(null, self.fromDB(model, mo, doc));
          });
        };
        self._insert(model, data, saveHandler);
      });
    }
  };

  overrideConnector._insert = function(model, data, cb) {
    if (!global.IMPORTED_TEST) {
      return originalConnector._insert(model, data, cb);
    } else {
      originalConnector._insert(model, data, function(err, rid, rrev) {
        if (err) return cb(err);
        cb(null, rid);
      });
    }
  };

  db.connector.save = overrideConnector.save;
  db.connector._insert = overrideConnector._insert;

  return db;
};

global.connectorCapabilities = {
  ilike: false,
  nilike: false,
  nestedProperty: true,
  supportPagination: false,
  ignoreUndefinedConditionValue: false,
  adhocSort: false,
  cloudantCompatible: false,
};

global.sinon = require('sinon');


test/describe.js
// Copyright IBM Corp. 2017. All Rights Reserved.
// Node module: loopback-connector-couchdb2
// This file is licensed under the Apache License 2.0.
// License text available at https://opensource.org/licenses/Apache-2.0

'use strict';

module.exports = process.env.CI ? describe.skip : describe;


lib/view.js
// Copyright IBM Corp. 2017,2019. All Rights Reserved.
// Node module: loopback-connector-couchdb2
// This file is licensed under the Apache License 2.0.
// License text available at https://opensource.org/licenses/Apache-2.0

'use strict';

var URL = require('url');
var assert = require('assert');
var util = require('util');
var _ = require('lodash');

module.exports = mixinView;

function mixinView(CouchDB) {
  var debug = require('debug')('loopback:connector:couchdb2:view');

  /**
   * Gets data at `/{db}/_design/{ddocName}/views/{viewName}`
   *
   * Example:
   * User has a view called `getModel` in design document /{db}/_design/model,
   * to query the view, user can call function:
   * ```
   * ds.viewDocs(model, getModel, {key: 'purchase'}, cb);
   * ```
   *
   * @param {String} ddocName The design doc name without {db}/_design/ prefix
   * @param {String} viewName The view name
   * @param {Object} options The CouchDB view filter
   * @callback
   * @param {Function} cb
   */
  CouchDB.prototype.viewDocs = function(ddocName, viewName, options, cb) {
    // omit options, e.g. ds.viewDocs(ddocName, viewName, cb);
    if (typeof options === 'function' && !cb) {
      cb = options;
      options = {};
    }
    debug('CouchDB2.prototype.view ddocName %s viewName %s options %s',
      ddocName, viewName, options);

    var self = this;
    var db = this.couchdb.use(self.getDbName(self));

    db.view(ddocName, viewName, options, cb);
  };

  /**
   * Return CouchDB database name
   * @param {Object} connector The CouchDB connector instance
   * @return {String} The database name
   */
  CouchDB.prototype.getDbName = function(connector) {
    var dbName = connector.settings.database || connector.settings.db ||
    getDbFromUrl(connector.settings.url);
    return dbName;
  };
};

/**
 * Parse url and return the database name if provided
 * @param {String} url The CouchDB connection url
 * @return {String} The database name parsed from url
 */
function getDbFromUrl(url) {
  var parsedUrl = URL.parse(url);
  if (parsedUrl.path && parsedUrl.path !== '/')
    return parsedUrl.path.split('/')[1];
  return '';
}


setup.sh
#!/bin/bash		
 		
 # Shell script to start the database and app services before running the tests.		
 		
 ## color codes		
 RED='\033[1;31m'		
 GREEN='\033[1;32m'		
 YELLOW='\033[1;33m'		
 CYAN='\033[1;36m'		
 PLAIN='\033[0m'		
 		
 ## variables		
 COUCH_CONTAINER="couch_c"
 COUCH_IMAGE="klaemo/couchdb"
 COUCH_IMAGE_TAG="latest"

 HOST=localhost		
 USER='admin'		
 PASSWORD='pass'		
 PORT=5984		
 DATABASE='testdb'		
 if [ "$1" ]; then		
     HOST=$1		
 fi		
 if [ "$2" ]; then		
     PORT=$2		
 fi		
 if [ "$2" ]; then		
     USER=$3		
 fi		
 if [ "$4" ]; then		
     PASSWORD=$4		
 fi		
 if [ "$5" ]; then		
     DATABASE=$5		
 fi		
 		
 ## check if docker exists		
 printf "\n${RED}>> Checking for docker${PLAIN} ${GREEN}...${PLAIN}"		
 docker -v > /dev/null 2>&1
 DOCKER_EXISTS=$?
 if [ "$DOCKER_EXISTS" -ne 0 ]; then		
     printf "\n\n${CYAN}Status: ${PLAIN}${RED}Docker not found. Terminating setup.${PLAIN}\n\n"		
     exit 1		
 fi	
 printf "\n${CYAN}Found docker. Moving on with the setup.${PLAIN}\n"

## cleaning up previous builds
printf "\n${RED}>> Finding old builds and cleaning up${PLAIN} ${GREEN}...${PLAIN}"
docker rm -f $COUCH_CONTAINER > /dev/null 2>&1
printf "\n${CYAN}Clean up complete.${PLAIN}\n"

## pull latest couch image
printf "\n${RED}>> Pulling latest couch image${PLAIN} ${GREEN}...${PLAIN}"
docker pull $COUCH_IMAGE:$COUCH_IMAGE_TAG > /dev/null 2>&1
printf "\n${CYAN}Image successfully built.${PLAIN}\n"

## run the couch container
printf "\n${RED}>> Starting the couch container${PLAIN} ${GREEN}...${PLAIN}"
CONTAINER_STATUS=$(docker run -d -e COUCHDB_USER=$USER -e COUCHDB_PASSWORD=$PASSWORD -p $PORT:5984 --name $COUCH_CONTAINER $COUCH_IMAGE:$COUCH_IMAGE_TAG 2>&1)
if [[ "$CONTAINER_STATUS" == *"Error"* ]]; then
    printf "\n\n${CYAN}Status: ${PLAIN}${RED}Error starting container. Terminating setup.${PLAIN}\n\n"
    exit 1
fi
printf "\n${CYAN}Container is up and running.${PLAIN}\n"
 		
 ## wait for couch service	
 OUTPUT=$?		
 TIMEOUT=120		
 TIME_PASSED=0		
 WAIT_STRING="."		
 		
 printf "\n${GREEN}Waiting for couch service to be up $WAIT_STRING${PLAIN}"		
 while [ "$OUTPUT" -ne 200 ] && [ "$TIMEOUT" -gt 0 ]		
     do		
         OUTPUT=$(curl -s -o /dev/null -w "%{http_code}" --request GET --url http://$USER:$PASSWORD@$HOST:$PORT/_all_dbs)		
         sleep 1s		
         TIMEOUT=$((TIMEOUT - 1))		
         TIME_PASSED=$((TIME_PASSED + 1))		
 		
         if [ "$TIME_PASSED" -eq 5 ]; then		
             printf "${GREEN}.${PLAIN}"		
            TIME_PASSED=0		
         fi		
    done		
		
if [ "$TIMEOUT" -le 0 ]; then		
    printf "\n\n${CYAN}Status: ${PLAIN}${RED}Failed to start Couch service. Terminating setup.${PLAIN}\n\n"		
    exit 1		
fi		
printf "\n${CYAN}Couch started.${PLAIN}\n"		

## create database		
printf "\n${RED}>> Creating database in Couch${PLAIN}"		
curl --request PUT --url http://$USER:$PASSWORD@$HOST:$PORT/$DATABASE > /dev/null 2>&1		
DB_OUTPUT=$?		
if [ "$DB_OUTPUT" -ne 0 ]; then		
    printf "\n\n${CYAN}Status: ${PLAIN}${RED}Database could not be created. Terminating setup.${PLAIN}\n\n"		
    exit 1		
fi		
printf "\n${CYAN}Database created succesfully.${PLAIN}\n"		

 ## set env variables for running test		
 printf "\n${RED}>> Setting env variables to run test${PLAIN} ${GREEN}...${PLAIN}"		
 export COUCHDB_URL=http://$USER:$PASSWORD@$HOST:$PORT		
 export COUCHDB_USERNAME=$USER		
 export COUCHDB_PASSWORD=$PASSWORD		
 export COUCHDB_PORT=$PORT		
 export COUCHDB_DATABASE=$DATABASE		
 export CI=true		
 printf "\n${CYAN}Env variables set.${PLAIN}\n"		
 		
 printf "\n${CYAN}Status: ${PLAIN}${GREEN}Set up completed successfully.${PLAIN}\n"		
 printf "\n${CYAN}Instance url: ${YELLOW}http://$USER:$PASSWORD@$HOST:$PORT/$DATABASE${PLAIN}\n"		
 printf "\n${CYAN}To run the test suite:${PLAIN} ${YELLOW}npm run mocha${PLAIN}\n\n"
